king_phisher/client/assistants/campaign.py

```diff
@@ -156,6 +278,61 @@ def _set_comboboxes(self):
 				combobox.add_attribute(renderer, 'text', 2)
 			combobox.set_model(rpc.get_tag_model(tag_table, model=model))
 
+	def _load_url_treeview_tsafe(self, hostname=None, refresh=False):
+		if refresh or not self._url_information['created']:
+			self._url_information['data'] = self.application.rpc.graphql_find_file('get_site_templates.graphql')
+			self._url_information['created'] = datetime.datetime.utcnow()
+		gui_utilities.glib_idle_add_once(self.gobjects['treeselection_url_selector'].unselect_all)
+		self._url_selected = None
+		gui_utilities.glib_idle_add_once(self._url_model.clear)
+		gui_utilities.glib_idle_add_once(self._domain_list_store.clear)
+
+		url_information = self._url_information['data']
+		if not url_information:
+			return
+
+		domains = []
+		self._url_classifiers = {}
+		for edge in url_information['siteTemplates']['edges']:
+			for page in edge['node']['metadata']['pages']:
+				if hostname:
+					if not edge['node']['hostname'].startswith(hostname):
+						continue
+				domains.append(edge['node']['hostname'])
+				self._url_classifiers[edge['node']['hostname']] = edge['node']['metadata']['classifiers']
+				row = _ModelNamedRow(
+					hostname=edge['node']['hostname'],
+					page=page,
+					url=self._build_url(edge['node']['hostname'], page),
+					classifiers='\n'.join(edge['node']['metadata']['classifiers']),
+					authors=', '.join(edge['node']['metadata']['authors']),
+					description=edge['node']['metadata']['description'].strip('\n'),
+					created=utilities.format_datetime(utilities.datetime_utc_to_local(edge['node']['created']))
+				)
+				gui_utilities.glib_idle_add_once(self._url_model.append, None, list(row))
+
+		# make domain list unique in case multiple pages are advertised for the domains
+		domains = [[domain] for domain in set(domains)]
+		gui_utilities.glib_idle_add_store_extend(self._domain_list_store, domains)
+
+	def _build_url(self, hostname, page):
+		if not hostname:
+			for address in self.config['server_config']['server.addresses']:
+				ip = ipaddress.ip_address(address['host'])
+				if not ip.is_unspecified and (ip.is_global or ip.is_private):
+					hostname = address['host']
+		if not hostname:
```

Comment:
Once you break out of the loop, this would be a great use case to use an [else clause](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops).

```
if not hostname:
    for address in self.config['server_config']['server.addresses']:
        ip = ipaddress.ip_address(address['host'])
        if not ip.is_unspecified and (ip.is_global or ip.is_private):
            hostname = address['host']
            break
    else:
        hostname = 'localhost'